lib/cppcheck.cpp

```diff
@@ -1824,7 +1824,7 @@ bool CppCheck::isUnusedFunctionCheckEnabled() const
     return (mSettings.jobs == 1 && mSettings.checks.isEnabled(Checks::unusedFunction));
 }
 
-void CppCheck::removeCtuInfoFiles(const std::map<std::string, std::size_t> &files)
+void CppCheck::removeCtuInfoFiles(const std::map<std::string, std::size_t> &files) const
```

Comment:
I have the idea that it's unfortunate to use "const" on a method that has side effects. Even though it's technically possible.

Reply:
I get your point, but we are missing several "technically safe" cases and we should not limit ourselves as unfortunately that is allowed by the standard and we already have cases where we don't take this into account. But the same issue exists in clang-tidy and I assume the GCC compiler warnings have the same problem.

We have a ticket about detecting functions that are not "safe" - see https://trac.cppcheck.net/ticket/11127.

Reply:
I think a better way would be to a properly protect the pointers with a "safe_ptr class which only offers `const T*` in a `const` context. I have a prototype.

Reply:
https://github.com/firewave/cppcheck/commit/76cc51edf391b83a809852a7e7e0f0c2f0f2b50b

Reply:
Turns out that be behavior I want to achieve matches the implementation of `std::optional`: https://en.cppreference.com/w/cpp/utility/optional/operator*. So maybe the check should be about that.

So let's take a look at the various cases:

`T *` - can be implemented with `std::optional<T*>`
`T * const` - can be implemented with `const std::optional<T*>`
`const T *` - object is always `const`; no issue
`const T * const` - object is always `const`; no issue

There's a conversion caveat though in the case a `nullptr` is being assigned to the pointer in the original code. In that case the optional musn't be set because `operator bool()` checks the content of itself and not the value of the pointer.